src/Application/Actions/Hooks/StripePaymentsUpdate.php

```diff
@@ -125,6 +127,16 @@ private function handleChargeSucceeded(Event $event, Response $response): Respon
             $donation->setChargeId($charge->id);
             $donation->setTransferId($charge->transfer);
 
+            /**
+             * @var Card|null $card
+             */
+            $card = $charge->payment_method_details?->card;
+            if ($card) {
+                /** @psalm-var value-of<Calculator::STRIPE_CARD_BRANDS> $brand */
+                $brand = $card->brand;
+                $this->donationRepository->deriveFees($donation, $brand, $card->country);
+            }
+
             $donation->setDonationStatus(DonationStatus::Collected);
             $donation->setCollectedAt(new \DateTimeImmutable("@{$charge->created}"));
```

Comment:
If this is for after CC23 so we're not trying to do it in a rush, I wonder if it would be worth going to the more rich domain model style and replacing all this, except getting the card from the stripe object, with a single function call on the Donation object. And the donation object could enforce the rule against setting any of them separately.

DonationRepository::deriveFees could be called from inside Donation, or imho even better it could be moved into donation. The only `$this` in deriveFees is to pass settings to the Calculator, but I think we'd be better off hard-coding all the settings inside the calculator, as changing settings isn't really any easier than editing the code, and we haven't tested the calculator with alternate settings anyway.

Reply:
(possibly isn't clear from the way github shows the comments what I meant by "all this". I mean all these statements that mutate the donation object:

```
$donation->setDonationStatus(DonationStatus::Collected);
$donation->setCollectedAt(new \DateTimeImmutable("@{$charge->created}"));
$this->donationRepository->deriveFees($donation, $brand, $card->country);
$donation->setTransferId($charge->transfer);
$donation->setChargeId($charge->id);
```

IMHO it's much better if you only have to make a single function call to mutate an object, and the object class can can enforce invariants before returning - can't have collectedAt without Collected status and vice versa, can't have Collected status without having calculated fees according the payment type etc.

Reply:
See https://github.com/thebiggive/matchbot/pull/730